halo-sigs/halo-gradle-plugin
# feat: add tasks for OpenAPI docs and api-client code generation

### What this PR does?
实现了两个 tasks，用于为插件项目生成 OpenAPI docs 和 API Client
- `generateOpenApiDocs`: 启动一个 Halo 作为 API Docs Server 并根据用户配置的 groupedApiMappings 来下载 Open API 的 json 文件到指定目录
- `generateApiClient` 根据 groupedApiMappings 指定的文件名称和其他配置来生成 API Client 到前端模块的项目中

**使用方式：**
在 build.gradle 中配置

```kotlin
haloPlugin {
    // 可以配置 configurationPropertiesFile 来指定插件需要的配置文件如应用市场插件
    // configurationPropertiesFile = file("${projectDir}/test.yaml")
    openApi {
        outputDir = file("$rootDir/api-docs/openapi/v3_0")
        // 用于定义 API 分组规则
         groupingRules {
            // 此名称为 group name，定义后 groupedApiMappings 中的 /v3/api-docs/ 后的名称需要与之相同，要避免与 halo 中已经存在的 group 相同避免生成后出现与插件无关的 API
            staticPagesExtensionV1alpha1Api {
                // 分组显示名称
                displayName = 'Extension API for Static Pages'
                // 分组的 API 规则
                pathsToMatch = ['/apis/staticpage.halo.run/v1alpha1/**']
            }
        }
        groupedApiMappings = [
                '/v3/api-docs/staticPagesExtensionV1alpha1Api': 'staticPagesExtensionV1alpha1Api.json'
        ]
        generator {
            // 默认配置可缺省
            outputDir = file("${projectDir}/console/src/api/generated")
            // 默认配置可缺省
            additionalProperties = [
                    useES6: true,
                    useSingleRequestParameter: true,
                    withSeparateModelsAndApi: true,
                    apiPackage: "api",
                    modelPackage: "models"
            ]
            // 默认配置可缺省
            typeMappings = [
                    set: "Array"
            ]
        }
    }
}
```

**how to test it?**
1. `./gradlew clean build -x check `
2. `./gradlew publishToMavenLocal`
3. 在插件项目中使用此插件 `id "run.halo.plugin.devtools" version "0.0.10-SNAPSHOT"`
4. 配置插件的 `settings.gradle` 中的 repositories 添加 `mavenLocal()`

```release-note
支持生成 API 文档（需升级 Gradle 至 8.3 及以上的版本）
```
